LION/models/PnP/drunet.py

```diff
@@ -0,0 +1,286 @@
+                for _ in range(model_parameters.n_blocks)
+            ],
+        )
+        self.up1 = torch.nn.Sequential(
+            upsample_convtranspose(
+                model_parameters.int_channels * 2,
+                model_parameters.int_channels,
+                bias_free=model_parameters.bias_free,
+            ),
+            *[
+                ResBlock(
+                    model_parameters.int_channels,
+                    kernel_size=model_parameters.kernel_size,
+                    bias_free=model_parameters.bias_free,
+                    act=self._act,
+                )
+                for _ in range(model_parameters.n_blocks)
+            ],
+        )
+
+        self.project = torch.nn.Conv2d(
+            model_parameters.int_channels,
+            model_parameters.out_channels,
+            kernel_size=(3, 3),
+            padding=(1, 1),
+            bias=not model_parameters.bias_free,
+        )
+
+    @staticmethod
+    def default_parameters():
+        return LIONParameter(
+            in_channels=1,
+            out_channels=1,
+            int_channels=64,
+            kernel_size=(3, 3),
+            n_blocks=4,
+            use_noise_level=False,
+            bias_free=True,
+            act="leaky_relu",
+            enforce_positivity=False,
+        )
+
+    @staticmethod
+    def cite(cite_format="MLA"):
+        if cite_format == "MLA":
+            print(
+                "Zhang, Kai, Yawei, Li, Wangmeng, Zuo, Lei, Zhang, Luc, Van Gool, Radu, Timofte."
+            )
+            print('"Plug-and-Play Image Restoration With Deep Denoiser Prior".')
+            print(
+                "\x1b[3mIEEE Transactions on Pattern Analysis and Machine Intelligence\x1b[0m"
+            )
+            print("44. 10(2022): 6360-6376.")
+        elif cite_format == "bib":
+            print("@article{zhang2022plug,")
+            print(
+                "author={Zhang, Kai and Li, Yawei and Zuo, Wangmeng and Zhang, Lei and Van Gool, Luc and Timofte, Radu},"
+            )
+            print(
+                "journal={IEEE Transactions on Pattern Analysis and Machine Intelligence},"
+            )
+            print("title={Plug-and-Play Image Restoration With Deep Denoiser Prior},")
+            print("year={2022},")
+            print("volume={44},")
+            print("number={10},")
+            print("pages={6360-6376},")
+            print("doi={10.1109/TPAMI.2021.3088914}")
+            print("}")
+        else:
+            raise ValueError(
+                f'`cite_format` "{cite_format}" is not understood, only "MLA" and "bib" are supported'
+            )
+
+    def forward(self, x0, noise_level: Optional[float] = None):
+        if self.model_parameters.use_noise_level:
+            assert (
+                isinstance(noise_level, float) and noise_level >= 0.0
+            ), "`noise_level` must be a non-negative float"
+            x0 = torch.cat((x0, noise_level * torch.ones_like(x0)), dim=1)
```

Comment:
Question: si in here you concatenate a number for the noise level? Is this like a "label" for the noise?

Reply:
Yes, I didn't use it in the benchmarking paper, but the paper that introduced DRUNet suggests this. Then you can train on multiple noise levels, while inputting both noisy images and the noise level, and you hope that it generalises to a spectrum of noise levels.